rfcs/20200721-extension-types.md

```diff
@@ -0,0 +1,1304 @@
+internally to implement several core TensorFlow data types, including
+`tf.SparseTensor`, `tf.RaggedTensor`, `tf.data.Dataset`, and
+`tf.StructuredTensor`.
+
+At a high level, types supported by this interface can be divided into two broad
+categories:
+
+* **_General data structures_**.
+  These types are handled by "generic" APIs whose behavior does not depend on
+  the value of each object (such as `tf.function`, `SavedModel`, and
+  `tf.while_loop`).
+
+* **_Tensor-like types_**, which specialize or extend tf.Tensor.
+  Values of these types have a `rank`, a `shape`, and usually a `dtype`.  In
+  addition to the "generic" APIs, these types can be handled by Tensor-specific
+  APIs (such as `tf.stack`, `tf.add`, and `tf.reduce_mean`).
+
+Examples of user-defined types that could defined or extended with this protocol
+include:
+
+**General data structures:**
+
+* `tfp.Distribution`: Encodes a statistical distribution.
+* `TensorDigraph`: Encodes the set of nodes and edges in a directed graph.
+* `DimensionAlignment`: Encodes a correspondence between two related dimensions
+  (e.g., between a `word` dimension and a `speaker` dimension).
+
+**Tensor-like types:**
+
+* `CSRSparseTensor`: A sparsely-encoded tensor that uses the Compressed Sparse
+  Row encoding.
+* `MaskedTensor`: Pairs a Tensor with a corresponding boolean mask, indicating
+  which values are valid, and automatically updates the mask as appropriate
+  when used with TensorFlow ops (such as `tf.add` or `tf.reduce_sum`).
+* `LabeledTensor`: Pairs a Tensor with a list of axis names, which can be used
+  for error detection and reporting, for specifying axes by name, and for
+  broadcasting.
+
+## User Benefit
+
+### Object-Oriented TensorFlow
+
+This proposal brings the benefits of Object-Oriented Programming to TensorFlow
+users, allowing them to define modular encapsulated data structures that
+interoperate with TensorFlow APIs.  This allows TensorFlow models to be defined
+at a higher level of abstraction.
+
+### Development outside the main TensorFlow repository
+
+Prior to this proposal, the only way to develop such data structures (e.g.
+`SparseTensor`) was to develop them inside the main TensorFlow code base.  This
+introduced significant barriers to rapid development, including slow release
+cycles, strong backwards compatibility constraints, and centralized API
+approval.  By allowing such data structures to be developed outside the main
+TensorFlow code base, we hope to make it much easier to experiment with new
+types and designs.  If general-purpose types are developed that become
+sufficiently mature, we may consider bringing them into the main TensorFlow code
+base.
+
+### TensorFlow APIs supported by user-defined types
+
+User-defined types that implement the interface proposed by this RFC will be
+supported by the following APIs:
+
+* **Keras**: User-defined types can be used as inputs and outputs for Keras
+  `Models` and `Layers`.
```

Comment:
How would a library like Keras support overrides of TensorFlow operations?

There are basically two alternatives:

1. It could allow the behavior of Keras objects (e.g., `Layer` instances) to be overriden _explicitly_ via some mechanism like `__tf_dispatch__`.
2. It could allow the behavior of Keras objects to be overriden _implicitly_ based on their underlying implementation in terms of TensorFlow operations.

The main challenge with (1) is that objects are (in general) stateful, so the override mechanism described here for functions would not suffice.

In contrast, (2) seems initially appealing. Keras is defined on arbitrary extension types "for free". The problem is that now Keras' implementation in terms of lower level TensorFlow function is exposed as part of its public API: if the underlying implementation of a Keras layer changes, it could break extension types, which might not implement a function, or worse might implement it in an incompatible way.

Reply:
This is an excellent question.  I would lean towards approach (2), but you're right to point out that this adds an element of design risk when using a Keras layer with an extension type.  Perhaps we can start with approach (2), but give appropriate warnings/disclaimers about the problem; and evaluate over the next few months whether this is a significant practical problem or not.

Reply:
To be clear, there are at least a few other options:

3. Keras doesn't (yet) support TF extension types (e.g., by always explicitly converting inputs with `tf.convert_to_tensor`).
4. Keras only supports TF extension types if an experimental flag is explicitly set. This flag might even be permanently required, if the Keras team doesn't want their implementation details to be considered public API.

I don't think shipping (2) and re-evaluating in a few months is a great option, because the costs of expanded API surface take a long time to become self-evident (i.e., at least a few major releases for users to start depending on a behavior and complain when it breaks). The point at which we notice that it's a problem could well be too late.

Reply:
There's also examples of this object-oriented dispatch question in the core TF APIs themselves (e.g. for the various tf.linalg.LinearOperator APIs)

The tricky thing is these may not provide any sort of canonical way to serialize/deserialize traced dispatched operations, because they can't be mapped to/from strings directly at the top-level api.